### DISCUSSION TITLE
Frequently encounter font issues when using the Binarykits Nuget package

### DISCUSSION THREAD

**Discussion by gayathrijosyula** [BinaryKits/ZPLUtility#172] [General]
We frequently encounter font issues when using the Binarykits Nuget package to convert ZPL to base64 content. When printing the labels, they do not appear in the correct alignment and font. A comparison with Labelary reveals the following differences in results. 

![image](https://github.com/BinaryKits/BinaryKits.Zpl/assets/121975085/faa8159a-d330-4e57-8b00-3659698ae5cf)

![image](https://github.com/BinaryKits/BinaryKits.Zpl/assets/121975085/3614cdc8-3aac-4067-acb1-9ed2f33f5046)

can you kindly suggest solution for this, also can you tell me if Binarykits and labelary have font hardcoded.

**Comment by primo-ppcg** [comment]
The font is not hardcoded, and can be adjusted by supplying a custom font loader to the drawer options ([example](https://github.com/BinaryKits/BinaryKits.Zpl#font-mapping)). If no font loader is supplied, Helvetica Semi-Condensed Bold will be used for Font 0, and DejaVu San Mono for all others, if installed.

I've corrected your ZPL by specifying a line count for the Field Block, and trailing it with a newline (this newline is necessary, or Labelary will not center it properly):

```zpl
^XA
^FT175,110^A0N,90^FD. Fullartons^FS
^FT175,230^A0N,90^FDRM13 7XH^FS
^FT750,220^A0N,85,40^FDR-7001^FS
^FT175,340^A0N,45^FDOrder 3100000787^FS
^FT750,480^A0N300,300^FB300,1,,C,^FD1\&^FS
^FT175,600^A0N,150^FD1/8^FS
^FT750,570^A0N,60^FD05/07/2023^FS
^FT750,600^A0N,30^FDLakeside Store (XP)^FS
^FO175,350^BY2,3.0^BCN,80,Y,N,N,N^FD3100000787^FS
^XZ
```

On my local system with default fonts:

![image](https://github.com/BinaryKits/BinaryKits.Zpl/assets/4907320/ef34c938-e2b5-414b-a13b-2a5981f7c2ae)

Labelary overlay in red.

The default line count for a field block should be 1, but it is overwritten to zero if provided an empty value. This is a bug that should be corrected (resolved with #174).

**Reply by gayathrijosyula** [reply]
Thank you,i am using Binarykits.Zplk.Viewer version 1.1.3 still experiencing issue. However, when i check the using a link, it doesn't seem to be present.

This is the Zpl string i am using : 

^XA^FT175,110^A0N,90^FD. Fullartons^FS^FT175,230^A0N,90^FDRM13 7XH^FS^FT750,220^A0N,85,40^FDR-7001^FS^FT175,340^A0N,45^FDOrder 3100000787^FS^FT750,480^A0N300,300^FB300,,,C,^FD1^FS^FT175,600^A0N,150^FD1/8^FS^FT750,570^A0N,60^FD05/07/2023^FS^FT750,600^A0N,30^FDLakeside Store (XP)^FS^FO175,350^BY2,3.0^BCN,80,Y,N,N,N^FD3100000787^FS^XZ

![image](https://github.com/BinaryKits/BinaryKits.Zpl/assets/121975085/39916bf2-cc0c-4de0-b275-8299eb3b883a)

Below result from Binarykits.Zplk.Viewer version 1.1.3

![image](https://github.com/BinaryKits/BinaryKits.Zpl/assets/121975085/08526475-aa53-49cb-8bd3-7ed36bf3fa2d)

**Reply by gayathrijosyula** [reply]
@primo-ppcg can you kindly help me with this please

**Comment by danzk** [comment]
The best approximation to Zebra Font 0 is Triumvirate Condensed Bold.

**Comment by gayathrijosyula** [comment]
Find below another ZPL string that contains miss alignment and font-related issues. Can you kindly help me in identifying the issue

Zpl string: 

^XA^FO130,375^GB909,0,9^FS^FO130,1575^GB909,0,9^FS^FO130,175^GB909,0,9^FS^FO130,175^GB0,1400,9^FS^FO1030,175^GB0,1400,9^FS^FO150,215^A0N,160^FDOVERNIGHT^FS^FO330,375^GB0,1200,3^FS^FO912,375^GB0,1200,3^FS^FO330,875^GB582,0,9^FS^FO330,1075^GB582,0,9^FS^FO330,975^GB97,0,9^FS^FT210,1555^A0B,70^FDOVERNIGHT^FS^FT264,1555^A0B,44^FDCON NUM : 066^FS^FT318,1555^A0B,44^FDDATE : 2023-05-24^FS^FO149,414^BY3,1^BCB,120,Y,N,N,N^FD0660001^FS^FT384,1555^A0B,44^FDMr DX Freight^FS^FT438,1555^A0B,44^FDaddr7^FS^FT492,1555^A0B,44^FDaddr7^FS^FT546,1555^A0B,44^FDaddr7^FS^FT600,1555^A0B,44^FDNewcastle^FS^FT654,1555^A0B,44^FD^FS^FT740,1555^A0B,44^FDITEM : 1 OF 1^FS^FT794,1555^A0B,44^FDWEIGHT : 80.0000 Kg^FS^FT848,1555^A0B,44^FDREF : 2305191432485^FS^FT902,1555^A0B,44^FDCARTON^FS^FT1002,1555^A0B,80^FDWickes Building Supplies ^FS^FT1002,725^A0B,44^FD AXS003658771 ^FS^FO427,875^GB291,0,200^FS^FT864,1051^A0B,170^FD^FS^FT420,725^A0B,80^FDEx : 54^FS^FO387,745^BY2,1^BCI,120,N,N,N,N^FD0660001^FS^FT730,725^A0B,300^FD38^FS^FT902,725^A0B,76^FDMK42 8NG^FS^XZ

![image](https://github.com/BinaryKits/BinaryKits.Zpl/assets/121975085/7016ed0f-e8b9-489d-a8e9-50ac19e554cc)

**Reply by primo-ppcg** [reply]
The only issue is that the Azure webhost does not have the default font installed. Run locally:

![image](https://github.com/BinaryKits/BinaryKits.Zpl/assets/4907320/8944c37b-b236-4a65-95fe-f467571a958a)

You can get even better results by extracting the proprietary Zebra Swis721 font from a printer, and using that.

It's also worth noting that while labelary is fairly accurate in most things, it is not precise.

**Reply by gayathrijosyula** [reply]
Can you provide guidance on the default font used in Azure web hosting at your end?.

**Reply by primo-ppcg** [reply]
I don't quite understand the question. The result shown above is from running the viewer from my local system, not on Azure.

**Reply by gayathrijosyula** [reply]
Thank you :) 
How are you able to obtain the displayed outcome when running the viewer locally, while I am unable to achieve the same result when running it on my own local machine? 

https://binarykits-zpl-viewer.azurewebsites.net/

**Reply by primo-ppcg** [reply]
I mean I'm actually hosting the Viewer on my local machine:

![image](https://github.com/BinaryKits/BinaryKits.Zpl/assets/4907320/7f1bee8f-909a-475f-be8f-38f8e73eee14)

You can clone the repository, and launch the Viewer.WebApi project from Visual Studio.

**Reply by gayathrijosyula** [reply]
I have incorporated this NuGet package into my local host code, but it is producing different results than expected.

![image](https://github.com/BinaryKits/BinaryKits.Zpl/assets/121975085/cbb27c63-0282-4c10-82dc-0c3a4b29c3d4)

**Reply by primo-ppcg** [reply]
You can check your local system that Helvetica Narrow Bold is installed. If not, you can install it, or specify a custom FontLoader in the DrawerOptions that are passed to the ZplElementDrawer.

**Comment by nazar-mykolyshyn** [comment]
Hello all, not sure if my question is related to this topic but I receive different versions of ZPL preview (I guess due to fonts) in https://binarykits-zpl-viewer.azurewebsites.net/ and local version running in Docker (https://hub.docker.com/r/yipingruan/binarykits-zpl using latest version (which is equal to dev-72 at the moment of posting). The azure one looks better

Do I need to supply some additional configuration to achieve the same result ? is docker image missing some fonts?
Thanks

**Comment by ivanjx** [comment]
anyone here know the exact font names used in zpl printers? the rendering below is using labelary viewer's initial script:

![image](https://github.com/user-attachments/assets/3fd6e190-b230-47b6-b3bf-bb8301aa197a)

**Reply by ivanjx** [reply]
i fixed font 0 by loading it with Triumvirate Condensed Bold as suggested by @danzk. the font A is still off though.

**Reply by danzk** [reply]
Even labelary.com doesn't render font A accurately.  You need a 5x9 pixel font like one of the following.

https://fontstruct.com/gallery/tag/371/5x9
